tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,479 @@
+
+```
+$ stack new todo-webapp
+```
+
+Then you can edit the cabal file to match the one provided in this tutorial. There's a couple of things you won't
+understand, but don't worry about it. The important thing is the
+`build-depends` section of your library, which should look like:
+
+```
+build-depends:         base                      >= 4.7     && < 5
+                     , bytestring                >= 0.9.1   && < 0.11
+                     , heist                     >= 0.14    && < 0.15
+                     , lens                      >= 4.12    && < 4.14
+                     , mtl                       >= 2       && < 3
+                     , snap                      >= 0.13    && < 0.15
+                     , snap-core                 >= 0.9     && < 0.10
+                     , snap-extras               >= 0.11    && < 0.12
+                     , snap-server               >= 0.9     && < 0.10
+                     , snap-loader-static        >= 0.9     && < 0.10
+                     , text                      >= 0.11    && < 1.3
+                     , xmlhtml                   >= 0.1     && < 0.3
+                     , HDBC                      >= 2.3     && <= 2.5
+                     , HDBC-sqlite3              == 2.3.*
+                     , interpolate               >= 0.1 && < 0.2
+```
+
+Those are the required libraries for the application to be built. Most of them
+belong to Snap's core like [`snap`](http://hackage.haskell.org/package/snap), [`snap-core`](http://hackage.haskell.org/package/snap-core) ,
+[`snap-server`](http://hackage.haskell.org/package/snap-server)
+and [`heist`](http://hackage.haskell.org/package/heist)(the HTML template library preferred by the Snap creators). There are
+others libraries like [`HDBC`](http://hackage.haskell.org/package/HDBC) and
+[`HDBC-sqlite3`](http://hackage.haskell.org/package/HDBC-sqlite3), which enable
+interaction with our database, and some others like [`text`](http://hackage.haskell.org/package/text)
+and [`interpolate`](http://hackage.haskell.org/package/interpolate) to
+allow us inmutable Unicode text manipulation and strings interpolation.
+
+## Coding our way out to the browser
+
+The heart of the Snap's snaplets infrastructure is state management. Based on that
+a good start for this application will be defining a data structure to hold the state,
+which includes the state of all snaplets used by our application. Let's define this type in a module called [Application](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Application.hs).
+
+```haskell
+data App = App
+  { _heist :: Snaplet (Heist App)
+  , _conn  :: Connection
+  }
+
+```
+This data structure contains a `_heist` field holding the
+[Heist Snaplet](http://snapforbeginners.com/chapters/heist.html)  and a `_conn`
+for the database connection. This is the right place to include all the Snaplets you
```

Comment:
We should use a connection pool or at least say that in a real application you want a connection pool.